### PR TITLE
Add message collectors

### PR DESCRIPTION
**Pull request by arynxd**
Update example

Add ability to remove event listeners

### CODE REVIEW

**Comment by arynxd** [comment]
The types were created in events/ because of circular imports. Creating it on a MessageChannel would be preferred but is impossible

**Comment by arynxd** [comment]
We also might want to close when we loose access to a channel, through permission update events?

**Review by TopiSenpai — APPROVED** [review]
looks good 
until I find a better solution for avoiding import cycles
damn I hate it

File: api/util/message_collector.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+func (c *MessageCollector) OnEvent(e interface{}) {
+	if event, ok := e.(*events.MessageCreateEvent); ok {
+		if !c.Filter(event.Message) {
+			return
+		}
+
+		c.Channel <- event.Message
+	} else if event, ok := e.(*events.GuildChannelDeleteEvent); ok && event.ChannelID == c.ChannelID {
+		c.Close()
+	} else if event, ok := e.(events.GuildVoiceLeaveEvent); ok && event.GuildID == c.GuildID {
```
Comment (Skye-31) [inline api/util/message_collector.go:59]: Was this meant to be Guild leave event, not guild *voice* leave event?

**Comment by arynxd** [comment]
yeah it is, I'll correct it when I get home.On 05/08/2021, 10:53 Skye ***@***.***> wrote:

   
  @Skye-31 commented on this pull request. 
   
  In api/util/message_collector.go: 
  > +	Close     func()
+	ChannelID api.Snowflake
+	GuildID   api.Snowflake
+}
+
+// OnEvent used to get events for the MessageCollector
+func (c *MessageCollector) OnEvent(e interface{}) {
+	if event, ok := e.(*events.MessageCreateEvent); ok {
+		if !c.Filter(event.Message) {
+			return
+		}
+
+		c.Channel <- event.Message
+	} else if event, ok := e.(*events.GuildChannelDeleteEvent); ok && event.ChannelID == c.ChannelID {
+		c.Close()
+	} else if event, ok := e.(events.GuildVoiceLeaveEvent); ok && event.GuildID == c.GuildID {
 
  Was this meant to be Guild leave event, not guild voice leave event?

**Comment by Skye-31** [comment]
This also needs a rebase

**Review by Skye-31 — APPROVED** [review]

**Pull request merged**